File: homere_bbb/scripts/sabertooth.py

import time, serial, struct


cmd_drive_forward_m1  = 0x00
cmd_drive_backward_m1 = 0x01
cmd_drive_forward_m2  = 0x04
cmd_drive_backward_m2 = 0x05

cmd_serial_timeout    = 0x0e
cmd_deadband          = 0x11

class SaberTooth:
    def __init__(self):
        self.device = '/dev/ttyS1'
        self.ser = serial.Serial(port=self.device, baudrate=9600)
        self.addr = 128
        self.set_serial_timeout(100)
        self.set_deadband(1)
        
    def set_serial_timeout(self, milisec):
        n = int(milisec/100)
        self.send_cmd(cmd_serial_timeout, n)

    def set_deadband(self, deadband):
        self.send_cmd(cmd_deadband, deadband)
        
        
    def send(self, m1, m2):
        ''' m1, m2: range -127:127 '''
        cmd, data = (cmd_drive_forward_m1, m1) if m1 >=0 else (cmd_drive_backward_m1, -m1)
        self.send_cmd(cmd, data)
        cmd, data = (cmd_drive_forward_m2, m2) if m2 >=0 else (cmd_drive_backward_m2, -m2)
        self.send_cmd(cmd, data)

    def send_cmd(self, cmd, data):
        checksum = self.addr+cmd+data & 0x7f
        s = struct.pack('cccc', chr(self.addr), chr(cmd), chr(data), chr(checksum))
        self.ser.write(s)
        
    def quit(self):
        self.send(0, 0)
        self.ser.close()





if __name__ == '__main__':
    mc = SaberTooth() 
    while True:
        mc.send(10, 10)
        time.sleep(0.05)

File: homere_bbb/src/sabertooth.cpp


#include "homere_bbb/sabertooth.h"

// stty -F /dev/ttyS5 sane

SaberTooth::SaberTooth():
  serial_("/dev/ttyS5", 9600),
  addr_(128) {
  
}

#define cmd_drive_forward_m1   0x00
#define cmd_drive_backward_m1  0x01
#define cmd_drive_forward_m2   0x04
#define cmd_drive_backward_m2  0x05
#define cmd_serial_timeout     0x0e
#define cmd_deadband           0x11

bool SaberTooth::init() {
  if (!serial_.init())
    {
      std::printf("Failed to initialize serial port\n");
      return false;
    }
  std::printf("initialized serial port\n");
  send(cmd_serial_timeout, 1);
  send(cmd_deadband, 1);
  return true;
}

void SaberTooth::send_drive(int m1, int m2) {
  if (m1 >= 0)
    send(cmd_drive_forward_m1, m1);
  else
    send(cmd_drive_backward_m1, -m1);
  if (m2 >= 0)
    send(cmd_drive_forward_m2, m2);
  else
    send(cmd_drive_backward_m2, -m2);
}


void SaberTooth::send(uint8_t cmd, uint8_t data) {
  uint8_t buf[4];
  buf[0] = addr_;
  buf[1] = cmd;
  buf[2] = data;
  buf[3] = (addr_ + cmd + data) & 0x7F;
  serial_.send_bytes(buf, 4);
  //std::printf("Sending %d %d %d %d\n", buf[0], buf[1], buf[2], buf[3]);
}

SaberTooth::~SaberTooth() {
  serial_.close();
}
